Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/KeyPackageSyncTool.kt

```diff
@@ -0,0 +1,138 @@
+    private val dayPackageSyncTool: DayPackageSyncTool,
+    private val hourPackageSyncTool: HourPackageSyncTool,
+    private val syncSettings: KeyPackageSyncSettings,
+    private val timeStamper: TimeStamper,
+    private val networkStateProvider: NetworkStateProvider
+) {
+
+    suspend fun syncKeyFiles(
+        wantedLocations: List<LocationCode> = listOf(LocationCode("EUR"))
+    ): Result {
+        cleanUpStaleLocation(wantedLocations)
+
+        val daySyncResult = runDaySync(wantedLocations)
+
+        val isMeteredConnection = networkStateProvider.networkState.first().isMeteredConnection
+        Timber.tag(TAG).d("Checking hour sync... (isMetered=%b)", isMeteredConnection)
+        val hourSyncResult = if (!isMeteredConnection || syncSettings.allowMeteredConnections.value) {
+            runHourSync(wantedLocations)
+        } else {
+            null
+        }
+
+        val availableKeys = keyCache.getAllCachedKeys()
+            .filter { it.info.isDownloadComplete }
+            .filter { (keyInfo, path) ->
+                path.exists().also {
+                    if (!it) Timber.tag(TAG).w("Missing keyfile for : %s", keyInfo)
+                }
+            }
+            .also { Timber.tag(TAG).i("Returning %d available keyfiles", it.size) }
+            .also { Timber.tag(TAG).d("Available keyfiles: %s", it.joinToString("\n")) }
+
+        val newKeys = mutableListOf<CachedKey>()
+        newKeys.addAll(daySyncResult.newPackages)
+        hourSyncResult?.let { newKeys.addAll(it.newPackages) }
+
+        return Result(
+            availableKeys = availableKeys,
+            newKeys = newKeys,
+            wasDaySyncSucccessful = daySyncResult.successful
+        )
+    }
+
+    private suspend fun cleanUpStaleLocation(acceptedLocations: List<LocationCode>) {
+        Timber.tag(TAG).d("Checking for stale location, acceptable is: %s", acceptedLocations)
+
+        val staleLocationData = keyCache.getAllCachedKeys()
+            .map { it.info }
+            .filter { !acceptedLocations.contains(it.location) }
+        if (staleLocationData.isNotEmpty()) {
+            Timber.tag(TAG).i("Deleting stale location data: %s", staleLocationData.joinToString("\n"))
+            keyCache.delete(staleLocationData)
+        } else {
+            Timber.tag(TAG).d("No stale location data exists.")
+        }
+    }
+
+    private suspend fun runDaySync(locations: List<LocationCode>): BaseKeyPackageSyncTool.SyncResult {
+        val lastDownload = syncSettings.lastDownloadDays.value
+        Timber.tag(TAG).d("Synchronizing available days (lastDownload=%s).", lastDownload)
+
+        syncSettings.lastDownloadDays.update {
+            KeyPackageSyncSettings.LastDownload(startedAt = timeStamper.nowUTC)
+        }
+
+        val successfulSync = dayPackageSyncTool.syncMissingDayPackages(
+            availableLocations = locations,
+            forceSync = lastDownload == null || !lastDownload.successful
```

Comment:
I still don't quite understand this. What does forceSync refer to from the tech spec?